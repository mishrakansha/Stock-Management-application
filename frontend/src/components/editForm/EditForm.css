.innerEditFormContainer {
  width: 100%;
  background-color: white;
  display: flex;
  justify-content: center;
  position: relative;
  flex-direction: column;
  box-shadow: 0px 0px 12px 1px #b3b3b3;
}
.innerEditFormContainer > div {
  color: white;
  padding: 1em;
  background: #323765;
  text-align: center;
}
.innerEditFormContainer > div > h1 {
  text-transform: uppercase;
  font-weight: 500;
}
#editForm {
  display: flex;
  flex-direction: column;
  row-gap: 1em;
  width: 100%;
}

.submitButtonContainer {
  display: flex;
  justify-content: center;
}
.submitButtonContainer > button {
  cursor: pointer;
  background: #323765;
}
.submitButtonContainer > button:hover {
  background: #323765;
  transform: scale(1.01);
}
.submitButtonContainer > button:disabled {
  cursor: not-allowed;
  pointer-events: all !important;
  color: white;
  background: #737373;
}
.flexInputBoxEditForm {
  width: 100%;
  justify-content: space-between;
  display: flex;
}
.flexInputBoxEditForm > div {
  width: 49%;
}

.flexInputBoxEditForm > div > div {
  width: 100%;
}
.flexEditFormErrorMessage {
  font-size: 0.7em;
  text-align: left;
  color: red;
}
@media screen and (max-width: 620px) {
  .flexInputBoxEditForm {
    flex-direction: column;
  }
  .flexInputBoxEditForm {
    flex-direction: column;
    gap: 1em;
  }
  .flexInputBoxEditForm > div {
    width: 100%;
  }
}
