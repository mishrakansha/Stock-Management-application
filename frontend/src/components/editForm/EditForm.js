import React, { Component } from "react";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";
import "./editForm.css";
import PropTypes from "prop-types";
import moment from "moment";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
class EditForm extends Component {
  constructor() {
    super();
    this.state = {
      itemName: "",
      quantity: "",
      price: "",
      description: "",
      date: "",
      manufacturingCompany: "",
      manufacturingCompanyError: false,
      itemNameError: false,
      quantityError: false,
      priceError: false,
      descriptionError: false,
      dateError: false,
      modified: false,
    };
  }
  async componentDidMount() {
    const { editFormId: id, getOneItem } = this.props;
    await getOneItem(id);
    const { oneItem: data } = this.props;
    const date = moment(data.date).format("yyyy-MM-DD");
    const { itemName, manufacturingCompany, price, quantity, description } =
      data;
    this.setState({
      itemName: itemName,
      quantity: quantity,
      price: price,
      description: description,
      date: date,
      manufacturingCompany: manufacturingCompany,
    });
  }
  handleSubmit = (event) => {
    event.preventDefault();
    const {
      itemName,
      quantity,
      price,
      description,
      date,
      manufacturingCompany,
    } = this.state;
    const data = {
      itemName: itemName,
      quantity: quantity,
      price: price,
      description: description,
      date: date,
      manufacturingCompany: manufacturingCompany,
    };
    const { editFormId: id, editItem, editFormPopUp } = this.props;
    editItem(id, data);
    editFormPopUp(false);
  };
  handelOnchange = async (event) => {
    const data = event.target.value;
    if (event.target.name === "itemName") {
      if (data.length > 2 && data.length <= 100) {
        await this.setState({ itemNameError: false });
      } else {
        await this.setState({ itemNameError: true });
      }
      await this.setState({ itemName: data });
    } else if (event.target.name === "quantity") {
      if (data >= 0 && data !== "") {
        await this.setState({ quantityError: false, quantity: parseInt(data) });
      } else {
        await this.setState({ quantityError: true, quantity: "" });
      }
    } else if (event.target.name === "description") {
      if (data.length > 0) {
        await this.setState({ descriptionError: false });
      } else {
        await this.setState({ descriptionError: true });
      }
      await this.setState({ description: data });
    } else if (event.target.name === "manufacturingCompany") {
      if (data.length > 0) {
        await this.setState({
          manufacturingCompanyError: false,
        });
      } else {
        await this.setState({ manufacturingCompanyError: true });
      }
      await this.setState({
        manufacturingCompany: data,
      });
    } else if (event.target.name === "price") {
      if (data > 0) {
        await this.setState({
          price: parseInt(data),
          priceError: false,
        });
      } else {
        await this.setState({
          price: "",
          priceError: true,
        });
      }
    } else if (event.target.name === "manufacturingCompany") {
      await this.setState({ manufacturingCompany: event.target.value });
    } else if (event.target.name === "date") {
      const inputtedDate = event.target.value;
      const today = new Date();
      const datenew = new Date(inputtedDate);
      if (today < datenew) {
        await this.setState({ dateError: true });
      } else {
        await this.setState({ dateError: false });
      }
      await this.setState({ date: data });
    }
    const { oneItem: unmodifiedData } = this.props;
    const originalDate = moment(unmodifiedData.date).format("yyyy-MM-DD");
    const {
      itemName: originalItemName,
      manufacturingCompany: originalManufecturingCompany,
      price: originalPrice,
      quantity: originalQuantity,
      description: originalDescription,
    } = unmodifiedData;
    const {
      itemName,
      quantity,
      price,
      description,
      date,
      manufacturingCompanyError,
      itemNameError,
      quantityError,
      priceError,
      descriptionError,
      dateError,
      manufacturingCompany,
    } = this.state;
    if (
      (originalItemName !== itemName ||
        originalQuantity !== quantity ||
        originalPrice !== price ||
        description !== originalDescription ||
        originalDate !== date ||
        originalManufecturingCompany !== manufacturingCompany) &&
      manufacturingCompanyError === false &&
      itemNameError === false &&
      quantityError === false &&
      priceError === false &&
      descriptionError === false &&
      dateError === false
    ) {
      this.setState({ modified: true });
    } else {
      this.setState({ modified: false });
    }
  };
  handelClose = () => {
    const { editFormPopUp } = this.props;
    editFormPopUp(false);
  };
  render() {
    const { isEditPopperOpen: open } = this.props;
    const {
      itemName,
      quantity,
      price,
      description,
      date,
      manufacturingCompany,
      manufacturingCompanyError,
      itemNameError,
      quantityError,
      priceError,
      descriptionError,
      dateError,
      modified,
    } = this.state;
    return (
      <Dialog open={open} onClose={this.handleClose} fullWidth={true}>
        <div className="innerEditFormContainer">
          <div>
            <h1>UPDATE ITEM</h1>
          </div>
        </div>
        <DialogContent>
          <form
            onSubmit={this.handleSubmit}
            method="post"
            action="localhost:5000/api/stock/additem"
            id="editForm"
          >
            <TextField
              required
              id="filled-basic itemName"
              name="itemName"
              value={itemName || ""}
              label="Item Name"
              variant="filled"
              InputProps={{
                style: {
                  height: "50px",
                },
              }}
              onChange={this.handelOnchange}
              type="text"
              style={{ width: "100%" }}
            />
            {itemNameError && (
              <div className="errorMessage">
                * Name of the item should contain atleast 3 character and atmost
                100 character
              </div>
            )}
            <div className="flexInputBoxEditForm">
              <div>
                <TextField
                  required
                  id="filled-basic"
                  name="quantity"
                  label="Quantity"
                  value={quantity || ""}
                  variant="filled"
                  InputProps={{
                    inputProps: {
                      min: 0,
                    },
                    style: {
                      height: "50px",
                    },
                  }}
                  onChange={this.handelOnchange}
                  type="number"
                />
                {quantityError && (
                  <div className="flexEditFormErrorMessage">
                    * Quantity should be greater than or equal to 0.
                  </div>
                )}
              </div>
              <div>
                <TextField
                  required
                  id="filled-basic"
                  name="price"
                  label="Price"
                  value={price || ""}
                  variant="filled"
                  InputProps={{
                    inputProps: {
                      min: 1,
                    },
                    style: {
                      height: "50px",
                    },
                  }}
                  onChange={this.handelOnchange}
                  type="number"
                />
                {priceError && (
                  <div className="flexEditFormErrorMessage">
                    * Price should be greater than 0.
                  </div>
                )}
              </div>
            </div>
            <TextField
              required
              id="filled-basic"
              name="manufacturingCompany"
              label="Manufacturing company"
              InputProps={{
                style: {
                  height: "50px",
                },
              }}
              variant="filled"
              onChange={this.handelOnchange}
              type="text"
              value={manufacturingCompany || ""}
              style={{ width: "100%" }}
            />
            {manufacturingCompanyError && (
              <div className="errorMessage">* this field is required.</div>
            )}
            <TextField
              required
              multiline
              rows={2}
              name="description"
              onChange={this.handelOnchange}
              variant="filled"
              value={description || ""}
              label="Description"
            />
            {descriptionError && (
              <div className="errorMessage">* this field is required.</div>
            )}
            <TextField
              id="datetime-local"
              label="Date Added On"
              type="date"
              variant="filled"
              value={date || ""}
              InputProps={{
                style: {
                  height: "50px",
                },
              }}
              name="date"
              onChange={this.handelOnchange}
              InputLabelProps={{
                shrink: true,
              }}
            />
            {dateError && (
              <div className="errorMessage">
                * date should not be greater than than today
              </div>
            )}
          </form>
        </DialogContent>
        <DialogActions className="submitButtonContainer">
          <Button
            fullWidth
            variant="contained"
            sx={{ mt: 3, mb: 2, width: "50%" }}
            className="closeButton"
            onClick={this.handelClose}
          >
            Close
          </Button>{" "}
          <Button
            disabled={!modified}
            title={!modified ? "kindly make any modification" : "Update"}
            onClick={this.handleSubmit}
            type="submit"
            onChange={this.handelOnchange}
            fullWidth
            variant="contained"
            sx={{
              mt: 3,
              mb: 2,
              width: "50%",
            }}
          >
            Update
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

EditForm.propTypes = {
  oneItem: PropTypes.object,
  isEditPopperOpen: PropTypes.bool,
  editFormId: PropTypes.string,
  editItem: PropTypes.func,
  getOneItem: PropTypes.func,
  editFormPopUp: PropTypes.func,
};
export default EditForm;
