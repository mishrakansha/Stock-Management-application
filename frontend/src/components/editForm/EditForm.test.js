import React from "react";
import { shallow } from "enzyme";
import EditForm from "./EditForm";
describe("EditForm", () => {
  const oneItem = {
    itemName: "Outlander Sport",
    quantity: 19,
    price: 4,
    description:
      "Maecenas tristique, est et tempus semper, est quam pharetra magna, ac consequat metus sapien ut nunc. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris viverra diam vitae quam. Suspendisse potenti.\n\nNullam porttitor lacus at turpis. Donec posuere metus vitae ipsum. Aliquam non mauris.\n\nMorbi non lectus. non, sodales sed, tincidunt eu, felis.\n\nFusce posuere felis sed lacus. Morbi sem mauris, laoreet ut, rhoncus aliquet, pulvinar sed, nisl. Nunc rhoncus dui vel sem.\n\nSed sagittis. Nam congue, risus semper porta volutpat, quam pede lobortis ligula, sit amet eleifend pede libero quis orci. Nullam molestie nibh in lectus.\n\nPellentesque at nulla. Suspendisse potenti. Cras in purus eu magna vulputate luctus.\n\nCum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vivamus vestibulum sagittis sapien. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.\n\nEtiam vel augue. Vestibulum rutrum rutrum neque. Aenean auctor gravida sem.\n\nPraesent id massa id nisl venenatis lacinia. Aenean sit amet justo. Morbi ut odio.\n\nCras mi pede, malesuada in, imperdiet et, commodo vulputate, justo. In blandit ultrices enim. Lorem ipsum dolor sit amet, consectetuer adipiscing elit.",
    date: "2022-10-18",
    manufacturingCompany: "Mitsubishi",
  };
  const getOneItem = () => {
    return oneItem;
  };
  const component = shallow(
    <EditForm
      getOneItem={getOneItem}
      oneItem={oneItem}
      isPopperOpen={true}
      editFormId={"u877e87t67etgy7t6ee"}
    />
  );
  it("Take a snapshot if do not added", () => {
    expect(component).toMatchSnapshot();
  });
  it("Test to render EditForm Heading", () => {
    expect(component.find(".innerEditFormContainer").render().text()).toEqual(
      "UPDATE ITEM"
    );
  });
  it("Test to render Update Button Text", () => {
    expect(component.find(".submitButtonContainer").text()).toEqual("Update");
  });
  it("Test to render the close button Button", () => {
    expect(component.find(".closeButton").text()).toEqual("Close");
  });
});
