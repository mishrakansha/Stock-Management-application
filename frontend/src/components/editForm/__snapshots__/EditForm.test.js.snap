// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm Take a snapshot if do not added 1`] = `
<ForwardRef(Dialog)
  fullWidth={true}
  open={true}
>
  <div
    className="innerEditFormContainer"
  >
    <div>
      <h1>
        UPDATE ITEM
      </h1>
    </div>
  </div>
  <ForwardRef(DialogContent)>
    <form
      action="localhost:5000/api/stock/additem"
      id="editForm"
      method="post"
      onSubmit={[Function]}
    >
      <ForwardRef(TextField)
        InputProps={
          Object {
            "style": Object {
              "height": "50px",
            },
          }
        }
        id="filled-basic itemName"
        label="Item Name"
        name="itemName"
        onChange={[Function]}
        required={true}
        style={
          Object {
            "width": "100%",
          }
        }
        type="text"
        value="Outlander Sport"
        variant="filled"
      />
      <div
        className="flexInputBoxEditForm"
      >
        <div>
          <ForwardRef(TextField)
            InputProps={
              Object {
                "inputProps": Object {
                  "min": 0,
                },
                "style": Object {
                  "height": "50px",
                },
              }
            }
            id="filled-basic"
            label="Quantity"
            min="1"
            name="quantity"
            onInput={[Function]}
            required={true}
            type="number"
            value={19}
            variant="filled"
          />
        </div>
        <div>
          <ForwardRef(TextField)
            InputProps={
              Object {
                "inputProps": Object {
                  "min": 1,
                },
                "style": Object {
                  "height": "50px",
                },
              }
            }
            id="filled-basic"
            label="Price"
            name="price"
            onChange={[Function]}
            required={true}
            type="number"
            value={4}
            variant="filled"
          />
        </div>
      </div>
      <ForwardRef(TextField)
        InputProps={
          Object {
            "style": Object {
              "height": "50px",
            },
          }
        }
        id="filled-basic"
        label="Manufacturing company"
        name="manufacturingCompany"
        onChange={[Function]}
        required={true}
        style={
          Object {
            "width": "100%",
          }
        }
        type="text"
        value="Mitsubishi"
        variant="filled"
      />
      <ForwardRef(TextField)
        label="Description"
        multiline={true}
        name="description"
        onChange={[Function]}
        required={true}
        rows={2}
        value="Maecenas tristique, est et tempus semper, est quam pharetra magna, ac consequat metus sapien ut nunc. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris viverra diam vitae quam. Suspendisse potenti.

Nullam porttitor lacus at turpis. Donec posuere metus vitae ipsum. Aliquam non mauris.

Morbi non lectus. non, sodales sed, tincidunt eu, felis.

Fusce posuere felis sed lacus. Morbi sem mauris, laoreet ut, rhoncus aliquet, pulvinar sed, nisl. Nunc rhoncus dui vel sem.

Sed sagittis. Nam congue, risus semper porta volutpat, quam pede lobortis ligula, sit amet eleifend pede libero quis orci. Nullam molestie nibh in lectus.

Pellentesque at nulla. Suspendisse potenti. Cras in purus eu magna vulputate luctus.

Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vivamus vestibulum sagittis sapien. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.

Etiam vel augue. Vestibulum rutrum rutrum neque. Aenean auctor gravida sem.

Praesent id massa id nisl venenatis lacinia. Aenean sit amet justo. Morbi ut odio.

Cras mi pede, malesuada in, imperdiet et, commodo vulputate, justo. In blandit ultrices enim. Lorem ipsum dolor sit amet, consectetuer adipiscing elit."
        variant="filled"
      />
      <ForwardRef(TextField)
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        InputProps={
          Object {
            "style": Object {
              "height": "50px",
            },
          }
        }
        id="datetime-local"
        label="Date Added On"
        name="date"
        onChange={[Function]}
        type="date"
        value="2022-10-18"
        variant="filled"
      />
      <div
        className="submitButtonContainer"
      >
        <ForwardRef(Button)
          disabled={true}
          fullWidth={true}
          onChange={[Function]}
          sx={
            Object {
              "mb": 2,
              "mt": 3,
              "width": "50%",
            }
          }
          type="submit"
          variant="contained"
        >
          Update
        </ForwardRef(Button)>
      </div>
    </form>
  </ForwardRef(DialogContent)>
  <ForwardRef(DialogActions)>
    <ForwardRef(Button)
      className="backButton"
      onClick={[Function]}
      sx={
        Object {
          ":hover": Object {
            "background": "rgb(50 55 101 / 4%)",
          },
          "color": " #323765",
        }
      }
    >
      Close
    </ForwardRef(Button)>
  </ForwardRef(DialogActions)>
</ForwardRef(Dialog)>
`;
